src/java.base/share/classes/jdk/internal/foreign/abi/SharedUtils.java

```diff
@@ -179,20 +187,22 @@ private static MethodHandle adaptUpcallForIMR(MethodHandle target, boolean dropR
         return target;
     }
 
-    public static UpcallStubFactory arrangeUpcallHelper(MethodType targetType, boolean isInMemoryReturn, boolean dropReturn,
+    public static UpcallStubFactory arrangeUpcallHelper(boolean isInMemoryReturn, boolean dropReturn,
                                                         ABIDescriptor abi, CallingSequence callingSequence) {
-        if (isInMemoryReturn) {
-            // simulate the adaptation to get the type
-            MethodHandle fakeTarget = MethodHandles.empty(targetType);
-            targetType = adaptUpcallForIMR(fakeTarget, dropReturn).type();
+        MethodType targetType = callingSequence.calleeMethodType();
```

Comment:
Previously this was using the targetType, which was derived from the FunctionDescriptor. Since that corresponds to the native side, it didn't contain the return buffer, and since t he return buffer isn't passed on to the user, that worked out find since the target type matches the type of the target method handle.

However, if we are capturing call state, we also need to pass on the capture state segment, so the target type has an additional leading MS parameter.

To make that work, I've switched to using the calleeMethodType here, which contains both the return buffer and capture state parameters, and then I filter out the return buffer below.